gax-java/gax/src/main/java/com/google/api/gax/logging/LoggingUtils.java

```diff
@@ -0,0 +1,467 @@
+import org.slf4j.helpers.MessageFormatter;
+import org.slf4j.helpers.NOPLogger;
+
+@InternalApi
+public class LoggingUtils {
+
+  private static final java.util.logging.Logger LOGGER =
+      java.util.logging.Logger.getLogger(LoggingUtils.class.getName());
+
+  private LoggingUtils() {}
+
+  public static Logger getLogger(Class<?> clazz) {
+    if (!isLoggingEnabled()) {
+      //  use SLF4j's NOP logger regardless of bindings
+      return LoggerFactory.getLogger(NOPLogger.class);
+    }
+
+    ILoggerFactory loggerFactory = LoggerFactory.getILoggerFactory();
+    if (loggerFactory != null && !(loggerFactory instanceof org.slf4j.helpers.NOPLoggerFactory)) {
+      // Use SLF4j binding when present
+      return LoggerFactory.getLogger(clazz);
+    }
+    // No SLF4j binding found, use JUL as fallback
+    Logger logger = new JulWrapperLogger(clazz.getName());
+    logger.info("No SLF4J providers were found, fall back to JUL.");
+    return logger;
+  }
+
+  public static boolean isLoggingEnabled() {
+    String enableLogging = System.getenv("GOOGLE_SDK_JAVA_LOGGING");
+    LOGGER.info("GOOGLE_SDK_JAVA_LOGGING=" + enableLogging); // log for debug now, remove it.
+    return "true".equalsIgnoreCase(enableLogging);
+  }
+
+  public static JsonObject mergeJsonObject(JsonObject jsonObject1, JsonObject jsonObject2) {
+    JsonObject mergedObject = jsonObject1.deepCopy();
+    jsonObject2.entrySet().forEach(entry -> mergedObject.add(entry.getKey(), entry.getValue()));
+    return mergedObject;
+  }
+
+  public static Level mapToJulLevel(org.slf4j.event.Level slf4jLevel) {
+    switch (slf4jLevel) {
+      case ERROR:
+        return Level.SEVERE;
+      case WARN:
+        return Level.WARNING;
+      case INFO:
+        return Level.INFO;
+      case DEBUG:
+        return Level.FINE;
+      case TRACE:
+        return Level.FINEST;
+      default:
+        return Level.INFO;
+    }
+  }
+
+  public static void logWithMDC(
+      Logger logger, org.slf4j.event.Level level, Map<String, String> contextMap, String message) {
+
+    if (logger instanceof JulWrapperLogger) {
+      // Simulate MDC behavior for JUL
+      LogRecord record = new LogRecord(mapToJulLevel(level), message);
+      // Add context map to the LogRecord
+      record.setParameters(new Object[] {contextMap});
+      ((JulWrapperLogger) logger).getJulLogger().log(record);
+      return;
+    }
+    contextMap.forEach(MDC::put);
+
+    switch (level) {
+      case TRACE:
+        logger.trace(message);
+        break;
+      case DEBUG:
+        logger.debug(message);
+        break;
+      case INFO:
+        logger.info(message);
+        break;
+      case WARN:
+        logger.warn(message);
+        break;
+      case ERROR:
+        logger.error(message);
+        break;
+      default:
+        logger.info(message);
+        // Default to INFO level
+    }
+
+    MDC.clear();
+  }
+
+  // JulWrapperLogger implementation
+  static class JulWrapperLogger implements Logger {
```

Comment:
This JulWrapper seems complicated and we have to override a bunch of methods without implementing them. I think it's OK that we don't include this in our first release.

Reply:
As discussed offline, will not add fallback option for first release.